src/proof/lrat/lrat_proof.cpp

```diff
@@ -0,0 +1,199 @@
+  }
+  formStream.close();
+
+  std::ofstream dratStream(dratFilename);
+  dratStream << dratBinary;
+  dratStream.close();
+
+  // TODO(aozdemir) Add invocation of DRAT trim, once I get CMake to bundle it
+  // into CVC4 correctly.
+  Unimplemented();
+
+  std::ifstream lratStream(lratFilename);
+  LRATProof lratProof(lratStream);
+  remove(formulaFilename);
+  remove(dratFilename);
+  remove(lratFilename);
+  *this = std::move(lratProof);
+}
+
+std::istream& operator>>(std::istream& in, prop::SatLiteral& l)
+{
+  long int i;
+  in >> i;
+  l = prop::SatLiteral(labs(i), i < 0);
+  return in;
+}
+
+// This parser is implemented to parse the textual RAT format found in
+// "Efficient Certified RAT Verification", by Cruz-Filipe et. All
+LRATProof::LRATProof(std::istream& textualProof)
+{
+  for (size_t line = 0;; ++line)
+  {
+    LRATInstruction instr;
+
+    // Read beginning of instruction. EOF indicates that we're done.
+    size_t clauseIdx;
+    textualProof >> clauseIdx;
+    if (textualProof.eof())
+    {
+      return;
+    }
+
+    // Read the first word of the instruction. A 'd' indicates deletion.
+    std::string first;
+    textualProof >> first;
+    Trace("pf::lrat") << "First word: " << first << std::endl;
+    Assert(textualProof.good());
+    if (first == "d")
+    {
+      instr.kind = lratDeletion;
+      instr.deletionData.idxOfClause = clauseIdx;
+      while (true)
+      {
+        ClauseIdx di;
+        textualProof >> di;
+        Assert(textualProof.good());
+        if (di == 0)
+        {
+          break;
+        }
+        instr.deletionData.clauses.push_back(di);
+      }
+      std::sort(instr.deletionData.clauses.begin(),
+                instr.deletionData.clauses.end());
+    }
+    else
+    {
+      instr.kind = lratAddition;
+      instr.additionData.idxOfClause = clauseIdx;
+
+      // We need to reparse the first word as a literal to read the clause
+      // we're parsing. It ends with a 0;
+      std::istringstream firstS(first);
+      SatLiteral lit;
+      firstS >> lit;
+      Trace("pf::lrat") << "First lit: " << lit << std::endl;
+      Assert(!firstS.fail(), "Couldn't parse first literal from addition line");
+      for (; lit != 0; textualProof >> lit)
+      {
+        Assert(textualProof.good());
+        SatLiteral l(lit.getSatVariable() - 1, lit.isNegated());
+        instr.additionData.clause.push_back(l);
+      }
+
+      // Now we read the AT UP trace. It ends at the first non-(+) #
+      long int i;
+      textualProof >> i;
+      for (; i > 0; textualProof >> i)
+      {
+        Assert(textualProof.good());
+        instr.additionData.atTrace.push_back(i);
+      }
+
+      // For each RAT hint... (each RAT hint starts with a (-).
+      for (; i<0; textualProof>> i)
+      {
+        Assert(textualProof.good());
+        // Create an entry in the RAT hint list
+        instr.additionData.resolvants.push_back(
+            make_pair(-i, std::vector<ClauseIdx>()));
```

Comment:
[Optional] As with the suggestion above, this could be written as `instr.additionData.resolvants.emplace_back(-i, std::vector<ClauseIdx>())`.

Reply:
Sounds good.